### ISSUE TITLE
Add method to check if component changed since last time step

### ISSUE THREAD

**Issue by osrf-migration** [osrf/gazebo_experimental#2]
**Original report ([archived issue](https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/gazebo_experimental/issues/2)) by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

----------------------------------------
Currently a system can query entities that have all required components. It gets a list of `EntityId` that match. This proposal is to add a way to see if a component on an entity has changed since the last update. It would mean :

* Add a method `Entity::IsDifferent(ComponentType)` which returns 0 if the component was not modified, or a non-zero value indicating if the component was created, deleted, or just changed.
* change `Entity::Component()` to return a pointer to a const object
* Add `Entity::ComponentMutable()` to return a pointer to a non-const component for writing and set an internal flag that the component was modified
* Add convenience methods using templates for the above.
* Supporting changes in `Manager`, `EntityComponentDatabase`, and `EntityQuery`.

The purpose is to enable a system to update its internal state only if there have been changes. For example, a physics system would only change the shape of an object if the shape changed, or stop simulating a body if the component describing it was removed.

Example

```
#!c++

void MySystemUpdate(const EntityQuery &_query)
{
  for (EntityId id : _query->EntityIds())
  {
    auto e = manager->Entity(id);
    switch (e.IsDifferent<components::WorldPose>())
      {
        case WAS_CREATED:
          // add thing
          break;
        case WAS_DELETED:
          // add thing
          break;
        case WAS_MODIFIED:
          // add thing
          break;
         default: break;
      }
      // ...
      if (needsChanging)
      {
        auto p = e.ComponentMutable<components::worldPose>();
        // ...  Do changes
      }
      else
      {
        auto p = e.Component<components::WorldPose>();
        // ... read only access to component
      }
  }
}
```

**Comment by osrf-migration** [comment]
**Original comment by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

----------------------------------------
Created and modified are in [pull request #7](https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/gazebo_experimental/pull-requests/7). I changed the name of `WasModified()` to `IsDifferent()`, and `WAS_CHANGED` to `WAS_MODIFIED`.

**Comment by osrf-migration** [comment]
**Original comment by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

----------------------------------------

* set assignee_account_id to "557058:684383ab-ac95-4859-a350-4a6f41a94a22"
* set assignee to "sloretz (Bitbucket: [sloretz](https://bitbucket.org/sloretz), GitHub: [sloretz](https://github.com/sloretz))"

**Comment by osrf-migration** [comment]
**Original comment by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

----------------------------------------

* Edited issue description
Changed methods to match names used in [pull request #7](https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/gazebo_experimental/pull-requests/7)

**Comment by osrf-migration** [comment]
**Original comment by Shane Loretz (Bitbucket: [Shane Loretz](https://bitbucket.org/%7B656e3311-aad9-45a1-aaf7-b0ee0e84b287%7D/), GitHub: [sloretz](https://github.com/sloretz)).**

----------------------------------------

* changed state from "new" to "resolved"

Fixed in [pull request #7](https://osrf-migration.github.io/osrf-archived-gh-pages/#!/osrf/gazebo_experimental/pull-requests/7)
